hooks/open-telemetry/pkg/otel.go

```diff
@@ -2,165 +2,52 @@ package otel
-	mu *sync.Mutex
-	ss *storedSpan // ss has been set as a pointer so it can be set to nil during cleanup
-}
-type storedSpan struct {
-	cancel func()
-	span   trace.Span
+	ctx context.Context
+	openfeature.UnimplementedHook
 }
 
 // NewHook return a reference to a new instance of the OpenTelemetry Hook
-func NewHook() *Hook {
+func NewHook(ctx context.Context) *Hook {
 	return &Hook{
-		tracerClient: &tracerClient{},
-	}
-}
-
-// Wait blocks until all spans have been closed
-func (h *Hook) Wait() {
-	h.wg.Wait()
-}
-
-// WithContext sets the parent context used when new spans are formed. When the parent context is closed so are all spans.
-// if no parent context is provided context.Background() is used internally
-func (h *Hook) WithContext(ctx context.Context) {
-	h.ctx = ctx
-}
-
-// Before creates the flag evaluations open-telemetry span and sets the FlagKey and ProviderName attributes
-func (h *Hook) Before(hookContext of.HookContext, hookHints of.HookHints) (*of.EvaluationContext, error) {
-	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
-	h.setup()
-	if _, ok := h.spans[key]; !ok {
-		h.spans[key] = &mutexWrapper{
-			mu: &sync.Mutex{},
-		}
-	}
-	h.spans[key].mu.Lock()
-	h.wg.Add(1)
-	ctx, span := h.tracerClient.tracer().Start(h.ctx, key)
-	ctx, cancel := context.WithCancel(ctx)
-	span.SetAttributes(
-		attribute.String(AttributeFlagKey, hookContext.FlagKey()),
-		attribute.String(AttributeProviderName, hookContext.ProviderMetadata().Name),
-	)
-	h.spans[key].ss = &storedSpan{
-		cancel: cancel,
-		span:   span,
+		ctx: ctx,
 	}
-	// this goroutine cleans up the span, if the associated context is closed then the stored
-	// span data is removed and the resource is unlocked. This context close can come from either the cancel() method
-	// or from the closing of the parent context outside the scope of the hook
-	go func() {
-		<-ctx.Done()
-		h.spans[key].ss = nil
-		h.spans[key].mu.Unlock()
-		h.wg.Done()
-	}()
-	return nil, nil
 }
 
-// After sets the EvaluatedVariant and EvaluatedValue on the evaluation specific span
-func (h *Hook) After(hookContext of.HookContext, flagEvaluationDetails of.InterfaceEvaluationDetails, hookHints of.HookHints) error {
-	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
-	mw, ok := h.spans[key]
-	if !ok {
-		return errors.New("no span stored for provided hook context")
-	}
-	if flagEvaluationDetails.ResolutionDetail.Variant != "" {
-		mw.ss.span.SetAttributes(
-			attribute.String(AttributeEvaluatedVariant, flagEvaluationDetails.ResolutionDetail.Variant),
-		)
+// After sets the feature_flag event and associated attributes on the span stored in the context
+func (h *Hook) After(hookContext openfeature.HookContext, flagEvaluationDetails openfeature.InterfaceEvaluationDetails, hookHints openfeature.HookHints) error {
+	if h.ctx == nil {
+		// if no context is set trace.SpanFromContext will return a noop span
```

Comment:
Is it better to simply return the no-op span then?

I'm trying to understand the benefits of handling this ourselves by returning `nul`, vs avoiding this check and returing a noop span.

Reply:
ive removed the nil check and allowed for the loop hook to be used